include/coro/task_container.hpp

```diff
@@ -172,38 +153,42 @@ class task_container
      * Grows each task container by the growth factor.
      * @return The position of the free index after growing.
      */
-    auto grow() -> task_position
+    auto grow() -> void
     {
         // Save an index at the current last item.
-        auto        last_pos = std::prev(m_task_indexes.end());
         std::size_t new_size = m_tasks.size() * m_growth_factor;
         for (std::size_t i = m_tasks.size(); i < new_size; ++i)
         {
-            m_task_indexes.emplace_back(i);
+            m_free_task_indices.emplace(i);
         }
         m_tasks.resize(new_size);
-        // Set the free pos to the item just after the previous last item.
-        return std::next(last_pos);
     }
 
     /**
-     * Interal GC call, expects the public function to lock.
+     * Internal GC call, expects the public function to lock.
      */
     auto gc_internal() -> std::size_t
     {
         std::size_t deleted{0};
-        if (!m_tasks_to_delete.empty())
+        auto   pos = std::begin(m_tasks_to_delete);
+        while (pos != std::end(m_tasks_to_delete))
         {
-            for (const auto& pos : m_tasks_to_delete)
+            // Skip tasks that are still running or have yet to start.
+            if (!m_tasks[*pos].is_ready())
             {
-                // Put the deleted position at the end of the free indexes list.
-                m_task_indexes.splice(m_task_indexes.end(), m_task_indexes, pos);
-                // Destroy the cleanup task and the user task.
-                m_tasks[*pos].destroy();
+                pos++;
+                continue;
             }
-            deleted = m_tasks_to_delete.size();
-            m_tasks_to_delete.clear();
+            // Destroy the cleanup task and the user task.
+            m_tasks[*pos].destroy();
+            // Put the deleted position at the end of the free indexes list.
+            m_free_task_indices.emplace(*pos);
+            // Remove index from tasks to delete
+            m_tasks_to_delete.erase(pos++);
+            // Indicate a task was deleted.
+            ++deleted;
         }
+        m_size.fetch_sub(deleted, std::memory_order::relaxed);
```

Comment:
I'm concerned about this change as well, the user task is "done" when `make_cleanup_task` adds itself to the `m_tasks_to_delete` list. This behavior will now make it seem like those tasks are still alive and running and won't drop until the user adds a new task and gc is invoked, or they manually invoke gc. Doesn't it still make sense to reduce the count at the end of `make_cleanup_task` to properly reflect how many user tasks are alive and running?

Reply:
I believe that would create a bug in the destructor and `garbage_collect_and_yield_until_empty` since both of those methods rely on calling `empty` to determine if all tasks have been garbage collected. It's possible that no tasks end up getting garbage collected when calling these methods since it would be possible for `m_size` to be 0 before `gc_internal` ever runs.

If the current behavior of `size` and `empty` is a concern, I could create a private size that tracks non-garbage collected tasks and running tasks. Program logic could use this without interfering with the public facing size that tracks only running tasks.

Reply:
You're absolutely right. I've been thinking on it and wondering if its worth it to track? Live vs Used?

Reply:
For my use case it's not helpful. I could image it might be useful for profiling or statistical purposes, but I don't know what else. I don't have a strong opinion on the matter, so I'll leave it to your judgement.

Reply:
The user's task is cleaned up so resources like sockets are free'ed, so you're right its probably not that useful. Thanks for talking through it, I'll get this merged.